### PR TITLE
Pong Game Beta, online basics working

### PR DESCRIPTION
**Pull request by Isioblast**
At the pr creation the game still need the following improvement :
- user id or nickname is used instead of entering a random username.
- saving the game result in db.
- handling when a user reload the page or quit the page.
- stop the game when someone score 12.
- improve ball behavior.
- add extra game mode.
- delete a game when both player are not ready for x seconds after the creation.
- adding possibility for two player to start a game without using the random queue.

### CODE REVIEW

**Review by Chicanoss — APPROVED** [review]
Nice work, keep going !

File: backend/src/pong-game/entities/pong-game.entity.ts [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+
+  @Field(() => Boolean)
+  S_Key = false
+
+  @Field(() => Boolean)
+  Up_Key = false
+
+  @Field(() => Boolean)
+  Down_Key = false
+}
```
Comment (kbarbry) [inline backend/src/pong-game/entities/pong-game.entity.ts:61]: Controls class present two times here and in player control input

File: backend/src/pong-game/entities/pong-game.entity.ts [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+  @Field(() => Int)
+  score = 0
+
+  @Field(() => Boolean)
+  presence: boolean
+
+  @Field(() => Controls)
+  controls = new Controls()
+}
+
```
Comment (kbarbry) [inline backend/src/pong-game/entities/pong-game.entity.ts:86]: All @ObjectType, @Field... etc decorators are made for graphql, why is it usefull everywhere on your class ? (all of them)

File: backend/src/pong-game/pong-game.resolver.spec.ts [hunk 3]
```diff
@@ -0,0 +1,18 @@
+import { Test, TestingModule } from '@nestjs/testing'
```
Comment (kbarbry) [inline backend/src/pong-game/pong-game.resolver.spec.ts:1]: Delete file

File: backend/src/pong-game/pong-game.resolver.ts [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+  })
+  pongData(@Args('gameId', { type: () => String }) gameId: string) {
+    console.log('Subscription: pongData:  gameId = ' + gameId)
+    return this.pubSub.asyncIterator(gameId)
+  }
+
+  //**************************************************//
+  //  MUTATION
+  //**************************************************//
+  @Mutation(() => String) //TODO delete this mutation
```
Comment (kbarbry) [inline backend/src/pong-game/pong-game.resolver.ts:42]: Todo now ? (Below as well)

File: backend/src/pong-game/pong-game.service.spec.ts [hunk 5]
```diff
@@ -0,0 +1,18 @@
+import { Test, TestingModule } from '@nestjs/testing'
```
Comment (kbarbry) [inline backend/src/pong-game/pong-game.service.spec.ts:1]: delete file

File: frontend/src/components/matchmaking/waiting-game.tsx [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+import { SubMatchmaking } from './subMatchmaking'
+
+type Props = {
+  username: string
+  setGameId: React.Dispatch<React.SetStateAction<string | null>>
+}
+
+export const WaitingGame: React.FC<Props> = (props: Props) => {
+  const [addPlayerToMatchmaking] = useMutation(addPlayerToMatchmakingQueue)
+
```
Comment (kbarbry) [inline frontend/src/components/matchmaking/waiting-game.tsx:13]: type useMutation

File: frontend/src/components/matchmaking/subMatchmaking.tsx [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+  setGameId: React.Dispatch<React.SetStateAction<string | null>>
+}
+
+export const SubMatchmaking: React.FC<Props> = (props: Props) => {
+  function listenMatchmaking(options: OnDataOptions<any>) {
+    let res = options.data.data.matchmakingNotification
+    props.setGameId(res)
+  }
+
+  useSubscription(matchmakingNotification, {
```
Comment (kbarbry) [inline frontend/src/components/matchmaking/subMatchmaking.tsx:16]: type useSubscription

File: frontend/src/components/game/subPong.tsx [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+  updateGameElement: (pongData: PongGame) => void
+}
+
+export const SubPong: React.FC<Props> = (props: Props) => {
+  function listenPongData(options: OnDataOptions<any>) {
+    let data = options.data.data.pongData as PongGame
+    props.updateGameElement(data)
+  }
+
+  useSubscription(pongDataSubscription, {
```
Comment (kbarbry) [inline frontend/src/components/game/subPong.tsx:17]: type useSubscription

File: frontend/src/components/game/subPong.tsx [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+    variables: { gameId: props.gameId },
+    onData: listenPongData,
+    onError: (error) => {
+      console.log('SubMatchmaking: onError: error = ' + JSON.stringify(error))
+    }
```
Comment (kbarbry) [inline frontend/src/components/game/subPong.tsx:22]: if error, juste console.log ? pas de throw ou de renvoi du matchmaking

File: frontend/src/components/game/controls-update.tsx [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+type Props = {
+  gameId: string
+  playerId: string
+  getControls: () => ControlsInput
+}
+
+export const ControlsUpdate: React.FC<Props> = (props: Props) => {
+  console.log('ControlsUpdate:')
+  const [updateInputs] = useMutation(updatePlayerInputs)
+
```
Comment (kbarbry) [inline frontend/src/components/game/controls-update.tsx:14]: type useMutation

File: frontend/src/components/game/Pong.tsx [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+  gameId: string
+  username: string
+  quitHandler: () => void
+}
+
+export const Pong: React.FC<Props> = (props) => {
+  console.log('Pong:')
+
+  const [sendReady, { data, loading, error }] = useMutation(readyForGame)
+
```
Comment (kbarbry) [inline frontend/src/components/game/Pong.tsx:15]: type useMutation

**Review by kbarbry — CHANGES_REQUESTED** [review]

File: backend/src/pong-game/entities/pong-game.entity.ts (hunk 1 above)
Comment (Isioblast) [reply backend/src/pong-game/entities/pong-game.entity.ts:61]: grpahql does not allox InputType() and ObjectType() for the same classe and for validation pipe.
Controls in game entity is used as part of the game class and control input is used to specified the controls send as argument for the resolver

File: backend/src/pong-game/entities/pong-game.entity.ts (hunk 2 above)
Comment (Isioblast) [reply backend/src/pong-game/entities/pong-game.entity.ts:86]: All the classe in game entity are returned to a user for the online game state, they all need to be specified as game type fo the graphql request, some @Field decorator will be deleted if they are not need in the request

**Pull request merged**